cmd/rpcdaemon/commands/eth_api.go

```diff
@@ -269,6 +274,54 @@ func (api *BaseAPI) headerByRPCNumber(number rpc.BlockNumber, tx kv.Tx) (*types.
 	return api._blockReader.Header(context.Background(), tx, h, n)
 }
 
+// checks the pruning state to see if we would hold information about this
+// block in state history or not.  Some strange issues arise getting account
+// history for blocks that have been pruned away giving nonce too low errors
+// etc. as red herrings
+func (api *BaseAPI) checkPruneHistory(tx kv.Tx, block uint64) error {
+	p, err := api.pruneMode(tx)
+	if err != nil {
+		return err
+	}
+	if p == nil {
+		// no prune info found
+		return nil
+	}
+	if p.History.Enabled() {
+		latest, err := api.blockByRPCNumber(rpc.LatestBlockNumber, tx)
+		if err != nil {
+			return err
+		}
+		prunedTo := p.History.PruneTo(latest.Number().Uint64())
+		if block < prunedTo {
+			return fmt.Errorf("history has been pruned for this block")
+		}
+	}
+
+	return nil
+}
+
+func (api *BaseAPI) pruneMode(tx kv.Tx) (*prune.Mode, error) {
+	api._pruneModeLock.Lock()
```

Comment:
Not RLock by intention?

Reply:
What version are we officially using now?  I thought 1.18?  If we're ok on 1.19 I can swap to an atomic pointer no problem.

The RLock thing was because I found a performance bottleneck on code called heavily when using RLock Vs a standard lock when looking at the Coherent Cache validation

Reply:
@hexoscott we using `uber/atomic` for 1.18 compatibility. can use atomic pointer from there.

Reply:
" I found a performance bottleneck on code called heavily when using RLock Vs a standard lock when looking at the Coherent Cache validation" - yes, now i remember. 
anyway I think atomic pointer is better for performance (because based on CompareAndSwap) and it's type-safe (can't forget lock/unlock).

Reply:
Perfect, thanks.  Preferable to swap the historyv3 over to an atomic pointer as well?